app/src/main/java/org/mozilla/fenix/components/metrics/AppStartupTelemetry.kt

```diff
@@ -143,4 +154,16 @@ class AppStartupTelemetry(private val metrics: MetricController) : LifecycleObse
         // Therefore we set the isMetricRecorded flag to false.
         isMetricRecordedSinceAppWasForegrounded = false
     }
+
+    /**
+     * We record metric on onPreDraw() since we need to wait for the application to draw the
+     * first frame. which will always happen after onResume. The hot startup time is a bit complicated
+     * because if we have onPreDraw listener attached, Activity will invoke onPreDraw() before onRestart.
+     * For this reason, we need to make sure that we remove the preDraw listener when activity is going
+     * in the background and reattach it during onRestart().
+     */
+    fun onPreDraw() {
+        appLaunchTimeMeasurement.onActivityOnPreDraw()
+        recordMetric()
```

Comment:
We shouldn't be doing such expensive computation in a pre-draw listener as this may cause us to drop frames: can we post it to run later or could this create invalid data? (I guess if we did it functionally and passed all the current values (assuming they're immutable) into the function as args, this wouldn't be a problem though – see why it's pre-emptively good to make things immutable 😄?).

Reply:
see https://github.com/mozilla-mobile/fenix/pull/14375#discussion_r486831613

Reply:
id say it pretty important to set data at this point since user can quit or background application anytime . once app is backgrounded, we have to start recording a fresh metric

Reply:
If we copy the data (which should be relatively simple with immutable objects – you just pass the reference rather than copying), we can send the metric slightly after this even if new data starts to get recorded.

Reply:
I commented in my review about this